src/main/java/org/candlepin/subscriptions/subscription/ExportSubscriptionListener.java

```diff
@@ -153,11 +181,83 @@ private void checkRbac(String appName, String identity) {
+
+    var filters = data.getFilters().entrySet();
+    var report = DbReportCriteria.builder().orgId(decodeIdentity);
+    for (var entry : filters) {
+      switch (entry.getKey().toLowerCase()) {
+        case "productid" -> report.productId(entry.getValue().toString());
+        case "usage" -> report.usage(Usage.fromString(entry.getValue().toString()));
+        case "category" ->
+            report.hypervisorReportCategory(
+                HypervisorReportCategory.valueOf(entry.getValue().toString()));
+        case "sla" -> report.serviceLevel(ServiceLevel.fromString(entry.getValue().toString()));
+        case "uom" -> report.metricId(entry.getValue().toString());
+        case "billingprovider" ->
+            report.billingProvider(BillingProvider.fromString(entry.getValue().toString()));
+        case "billingaccountid" -> report.billingAccountId(entry.getValue().toString());
+        default -> log.warn("Filter {} isn't supported currently", entry.getKey());
+      }
+    }
+    return report.build();
+  }
+
+  private SubscriptionsExport transformSubscriptionForExport(Stream<Subscription> subscription) {
+
+    var subscriptionExport = new SubscriptionsExport();
+
+    var subscriptionsStream =
+        subscription
+            .map(
+                subscription1 -> {
+                  var subscriptionJson = new org.candlepin.subscriptions.json.Subscription();
+                  var summaries = new ArrayList<SubscriptionSummary>();
+                  subscriptionJson.setOrgId(subscription1.getOrgId());
+                  subscriptionJson.setSku(subscription1.getOffering().getSku());
+                  subscriptionJson.setUsage(subscription1.getOffering().getUsage().getValue());
+                  subscriptionJson.setProductName(subscription1.getOffering().getProductName());
+                  subscriptionJson.setServiceLevel(
+                      subscription1.getOffering().getServiceLevel().getValue());
+
+                  for (var entry : subscription1.getSubscriptionMeasurements().entrySet()) {
+                    var summary = new SubscriptionSummary();
+                    summary.setMeasurementType(entry.getKey().getMeasurementType());
+                    summary.setCapacity(entry.getValue());
+                    summary.setMetricId(entry.getKey().getMetricId());
+                    summary.setSubscriptionNumber(subscription1.getSubscriptionNumber());
+                    summary.setQuantity(subscription1.getQuantity());
+                    summary.setBeginning(subscription1.getStartDate());
+                    summary.setEnding(subscription1.getEndDate());
+                    summaries.add(summary);
+                  }
+                  subscriptionJson.setSubscriptionSummaries(summaries);
+                  return List.of(subscriptionJson);
+                })
+            .flatMap(Collection::stream);
+
+    subscriptionExport.setSubscriptions(subscriptionsStream.toList());
```

Comment:
Will the toList() load the full results into memory?

Reply:
Yes, I created a card -> https://issues.redhat.com/browse/SWATCH-2186, as it would have been messy to try to optimize in this card, so I try to leverage stream as much as I can before returning it in memory, @kahowell and I decide it would be best to get the changes in and than optimize while adding the implementation for csv files